src/corona_hakab_model/generation/matrix_generator.py

```diff
@@ -172,194 +194,133 @@ def _create_random_clustered_circles_matrix(self, con_type: ConnectionTypes, cir
+            for node, num_connections in zip(islice(nodes, initial_con_amount, None), connections_amounts):
                 # selects the first node to attach to randomly
-                rand_node = choice(list(inserted_nodes))
-                inserted_nodes.remove(rand_node)
-
-                # adds a connection between the nodes
-                connections[node.index].append(rand_node.index)
-                connections[rand_node.index].append(node.index)
-
-                # todo change this to use p, and not only p = 1
-                # randomly choose the rest of the connections from rand_node connections.
-                nodes_to_return = []
-                for _ in range(connections_amount - 1):
-                    # randomly choose a node from rand_node connections
-                    new_rand = rand_node.pop_random()
-                    nodes_to_return.append(new_rand)
-                    Node.connect(node, new_rand)
-
-                    # add connection to connections list
-                    connections[node.index].append(new_rand.index)
-                    connections[new_rand.index].append(node.index)
-                # connect current node with rand node. note that this only happens here to not pick yourself later on
-                Node.connect(node, rand_node)
-                # return the popped nodes back to rand_node connections, and rand node back to already inserted list
-                rand_node.add_connections(nodes_to_return)
-                inserted_nodes.add(rand_node)
-                inserted_nodes.add(node)
+                first_connection = choice(list(connected_nodes))
+                # fill connections other than first_connection
+                while len(node.connected) < num_connections - 1:
+                    if random() < con_type_data.triad_p:
+                        # close the triad with a node from first_connection's connections
+                        possible_nodes = first_connection.connected
+                    else:
+                        # connect with a node, which is not first connection
+                        possible_nodes = connected_nodes
 
-        # adding connections between all super small circles
-        self._add_small_circle_connections(super_small_circles_combined, connections, total_connections_float)
+                    # prevent connecting a connected node
+                    if len(possible_nodes) - len(node.connected) > 100:  # TODO 100 is made up. should be optimized
+                        possible_list = list(possible_nodes)
+                        random_friend = choice(possible_list)
+                        while random_friend in node.connected:
+                            random_friend = choice(possible_list)
+                    else:
+                        possible_nodes = possible_nodes.difference(node.connected).difference({first_connection})
+                        random_friend = choice(list(possible_nodes))
```

Comment:
is there any chance that possible_nodes is gonna be empty?

Reply:
Shouldn't be. That's why it now begins with initiating more than 1 node (was this way at work and school already, just added it to other too).